augur/export_v2.py

```diff
@@ -1205,7 +1232,8 @@ def run(args):
     set_filters(data_json, config)
 
     # set tree structure
-    data_json["tree"] = convert_tree_to_json_structure(T.root, node_attrs, node_div(T, node_attrs))
+    trees_json = [convert_tree_to_json_structure(tree.root, node_attrs, node_div(tree, node_attrs)) for tree in trees]
+    data_json["tree"] = trees_json[0] if len(trees_json)==1 else trees_json
```

Comment:
Why not always use a list of trees here, even if it's a one-item list? This would simplify internal logic by avoiding the need to check data type on every reference.

Having multiple trees under the singular key `tree` is also a bit confusing. Would it be too much to put this in a new plural key `trees` that's more self-descriptive?

Auspice could continue to support a single tree object for backwards compatibility.

Reply:
In terms of the final dataset JSON, I'd prefer to keep the single-tree export as it currently is to remain consistent with output many users have become familiar with (and many people have written scripts to expect, as it's the case in nearly all situations). 

For the multiple trees export it may help to conceptually think of these as subtrees. The schema has supported an array of (sub)trees since #851 and while `trees` may be nicer in some regards I don't think it's worth changing the schema here.

> Why not always use a list of trees here, even if it's a one-item list? This would simplify internal logic by avoiding the need to check data type on every reference.

In terms of internal implementation, we could use an array which would reduce the conditional logic in the code, and then essentially run line 1236 at the very end to export a single tree object where applicable. I felt this changing of types was more confusing to developers, but 🤷

Reply:
Gotcha, makes sense to keep the final JSON as-is.

For internal implementation, I find it confusing when a variable takes on two types especially in an untyped language. Thinking in terms of subtrees, a more concrete suggestion would be to rename internal variables to always assume a list of `subtrees` instead of `tree` which is confusingly one or multiple trees. Then at the very end export dynamically based on if it's a singular tree or multiple subtrees. Does that make sense?

I've not worked much on this part of the code, so take my comments with a grain of salt!

Reply:
I can appreciate that maintaining multiple types without any type-hints (which AFAIK is the best python can do) requires developer overhead and won't scale. I chose this pattern as it matches the pattern already in the codebase: https://github.com/nextstrain/augur/blob/03ed408bbd84eff4d27f0c3b07757a76b5171045/augur/export_v2.py#L113-L117 but that doesn't prevent us from improving.

> a more concrete suggestion would be to rename internal variables to always assume a list of subtrees

I think this would be more confusing if we continue to use `data_json['tree']` (unsure if that's what you're suggesting), but I've implemented this using the new variable `trees`. Could you check out [c40b821](https://github.com/nextstrain/augur/pull/1450/commits/c40b821f9b2ee22cbf39ecf11232ce160d09f17a)?

Reply:
I see, sorry I should've tried this out first. That doesn't seem any better since we'd still have to keep conditions elsewhere for `data_json['tree']` such as

https://github.com/nextstrain/augur/blob/c40b821f9b2ee22cbf39ecf11232ce160d09f17a/augur/validate_export.py#L21-L25

I'm fine with dropping c40b821f9b2ee22cbf39ecf11232ce160d09f17a and keeping how you had it before.

Reply:
> without any type-hints (which AFAIK is the best python can do)

We could use something like the following [TypedDict](https://docs.python.org/3.8/library/typing.html#typing.TypedDict) for `data_json`, but enforcement (#1246) and usefulness across text editors is another story.

```py
Tree = Dict[str, Any]

class DataJson(TypedDict):
    tree: Tree or List[Tree]
    meta: ...
```